### PR TITLE
11 feature use auth hook

### PR DESCRIPTION
**Pull request by Joepock123**
- Create custom hook which listens/subscribes to auth changes using a useEffect, updates the firebaseUser state which is managed using the useContext hook, and then returns the firebaseuser
- This user is either null or an object
- Depending on the state the app.tsx renders different routers
- The nav also render different links

### CODE REVIEW

File: src/components/pages/auth/sign_up/SignUp.tsx [hunk 1]
```diff
@@ -1,29 +1,51 @@
...
     setUserData(shallowCloneUserData);
   };
 
+  // See if we can refactor this to have one change handler
+  const handleChangeCheckbox = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const property = event.target.name;
+    const isCheckedValue = event.target.checked;
+
+    const shallowCloneUserData = { ...userData, [property]: isCheckedValue };
+    setUserData(shallowCloneUserData);
```
Comment (VatsKan) [inline src/components/pages/auth/sign_up/SignUp.tsx:40]: did you manage to get the true-false the right way round in the end?

File: src/components/shared_components/buttons/hamburger_menu/HamburgerMenu.tsx [hunk 2]
```diff
@@ -1,23 +1,49 @@
...
+};
+
+// Notes on withRouter
+// withRouter will pass updated match, location,
+// and history props to the wrapped component whenever it renders.
+// It connects components that are not wrapped in 'route'
+// with the router
 
-export default TempNav;
+export default withRouter(TempNav);
```
Comment (VatsKan) [inline src/components/shared_components/buttons/hamburger_menu/HamburgerMenu.tsx:49]: dont need withRouter when using react-router-dom hooks (its the old way of doing things). better to use useHistory which contains all location match history info if i remember correctly

File: src/components/pages/auth/sign_up/SignUp.tsx [hunk 3]
```diff
@@ -1,29 +1,51 @@
...
 
   const history = useHistory();
 
+  // Note: should we initialize auth once and then use variable like so:
+  // const auth = firebase.auth()
+  // Then we can use auth.method instead
+  // Done that way in FB tutorial
+
   const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
     const property = event.target.name;
```
Comment (VatsKan) [inline src/components/pages/auth/sign_up/SignUp.tsx:26]: handleChange and handleChangeCheckbox have very similar code and can probably be merged into a single function (with an if statement) if we wanted to refactor at some point

Comment (VatsKan) [reply src/components/pages/auth/sign_up/SignUp.tsx:26]: perhaps can also be extracted out into utils folder for use in all forms

File: src/firebase/Auth.jsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import React, { useState, createContext } from 'react';
```
Comment (VatsKan) [inline src/firebase/Auth.jsx:1]: why make this .jsx file?

File: src/utils/use-auth.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  useEffect(() => {
+    const unsubscribe = firebase.auth().onAuthStateChanged((firebaseUser) => {
+      if (firebaseUser) {
+        setUser(firebaseUser);
+      } else {
+        setUser(null);
+      }
+    });
+    // I don't know why we need to unmount this
+    return () => unsubscribe();
```
Comment (VatsKan) [inline src/utils/use-auth.tsx:26]: will be interesting to look into this why it needs unmounting

File: src/firebase/Auth.jsx (hunk 4 above)
Comment (Joepock123) [reply src/firebase/Auth.jsx:1]: tee hee 
We'll change it back. Promise

File: src/utils/use-auth.tsx (hunk 5 above)
Comment (Joepock123) [reply src/utils/use-auth.tsx:26]: Agreed

**Pull request merged**